### PR TITLE
Feature/issue 863 compile time

### PR DESCRIPTION
**Pull request by rok-cesnovar**
#### Submisison Checklist

- [x] Run tests: `./runCmdStanTests.py src/test`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

Addresses a part of #863 by re-doing some parts of #769 

I checked that #763 and #765 are still fixed after this.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Rok Češnovar

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: make/program [hunk 1]
```diff
@@ -43,12 +45,15 @@ endif
...
+%.d: %.hpp
+
 .PRECIOUS: %.hpp
-%$(EXE) : %.hpp %.d $(CMDSTAN_MAIN_O) $(LIBSUNDIALS) $(MPI_TARGETS) $(TBB_TARGETS)
+%$(EXE) : %.hpp $(CMDSTAN_MAIN_O) $(LIBSUNDIALS) $(MPI_TARGETS) $(TBB_TARGETS)
 	@echo ''
 	@echo '--- Compiling, linking C++ code ---'
 	$(COMPILE.cpp) $(CXXFLAGS_PROGRAM) -x c++ -o $(subst  \,/,$*).o $(subst \,/,$<)
 	$(LINK.cpp) $(subst \,/,$*.o) $(CMDSTAN_MAIN_O) $(LDLIBS) $(LIBSUNDIALS) $(MPI_TARGETS) $(TBB_TARGETS) $(subst \,/,$(OUTPUT_OPTION))
+	$(RM) $(subst  \,/,$*).o
```
Comment (rok-cesnovar) [inline make/program:56]: This is here so that one case of #763 is not an issue:

- compile
- remove hpp
- change model
- recompile

**Comment by rok-cesnovar** [comment]
I would appreciate if someone with a mac can verify.
Do: 
make build
make examples/bernoulli/bernoulli

touch examples/bernoulli/bernoulli.stan
time make examples/bernoulli/bernoulli

and repeat the last two a few times.

**Comment by rok-cesnovar** [comment]
Not ready yet, tests fail to compile. Will reopen when ready.

**Comment by rok-cesnovar** [comment]
@serban-nicusor-toptal  the Windows issue stil persists. Is it a specific worker? Can we shut it down?

**Comment by mitzimorris** [comment]
> I would appreciate if someone with a mac can verify.
> Do:
> make build
> make examples/bernoulli/bernoulli
> 
> touch examples/bernoulli/bernoulli.stan
> time make examples/bernoulli/bernoulli
> 
> and repeat the last two a few times and see if there is a difference between develop and this PR.

@rok-cesnovar I can test this on both Mac High Sierra and Catalina - should I do this now?

for these speedups, does it matter if threading is on or not?

**Comment by rok-cesnovar** [comment]
Yes, this is ready to test. On Clang I see ~20% speedup of this PR vs develop.
STAN_THREADS should not matter- The baseline will probably be a bit higher (more stuff to compile) but the speedup should be there.

**Comment by rok-cesnovar** [comment]
It seems I will need to address test compiling, this may have broken it for clang. Regular models still compile though.

**Comment by rok-cesnovar** [comment]
Hey @serban-nicusor-toptal, is the gelman windows machine offline or still has the make issue?
There is a backlog of windows jobs. No worries, if the machine still has issues.

**Comment by serban-nicusor-toptal** [comment]
Hey, I think it's fixed now, gonna put it back online. Not sure why but the PATH was totally broken, managed to replicate what's on the other machine.

**Comment by mitzimorris** [comment]
I have been comparing branches `develop`, `feature/issue-863-compile-time` -
in addition to testing model compilation, we also need to check that changes to stan algorithms trigger the proper recompilation, not just for Stan programs but also for the unit and interface tests as well.

for Stan model compilation, the test procedure is:

```
make clean-all
make build

time make examples/bernoulli/bernoulli

# test changes to model trigger recompile

touch examples/bernoulli/bernoulli.stan
time make examples/bernoulli/bernoulli

# repeat touch / time / touch / time,  ...

# test changes to stan algorithms trigger recompile

touch stan/src/stan/mcmc/var_adaptation.hpp
time make examples/bernoulli/bernoulli

# repeat touch / time / touch / time,  ...
```

this works.

however, changes to core stan should also trigger rebuilds in the unit test (I think) -
this gave me different behavoirs between develop and branch 863 -

```
touch stan/src/stan/mcmc/var_adaptation.hpp
time ./runCmdStanTests.py src/test/interface/metric_test.cpp
```

dev picked up changes and recompiled `metric_test.o`.  this branch didn't.

**Comment by rok-cesnovar** [comment]
@mitzimorris good catch! Thanks! Should be fixed now.

**Comment by rok-cesnovar** [comment]
Also, have you seen any speedup in compile time on mac?

**Comment by mitzimorris** [comment]
> Also, have you seen any speedup in compile time on mac?
yes, definite speedup.
will test latest and report numbers.

**Comment by mitzimorris** [comment]
Compile time comparison on Catalina - 10 runs, command:
```
> for x in {1..10}; do touch examples/bernoulli/bernoulli.stan ; time (make examples/bernoulli/bernoulli >& /dev/null); done >& times.txt
```

this branch, average time over 10 runs:
real: 9.93s
sys: 0.6s
user: 9.29s

develop branch:
real: 13.11s
sys: 1.04s
user: 12.01s

**Comment by mitzimorris** [comment]
timing tests, Mac High Sierra, machine MacBook Pro, 3.1 GHz Intel Core i7, 16MB Ram

same drill as above

this branch:
real=5.88
sys=0.43
user=6.69

develop:
real=9.05
sys=0.68
user=8.44

nice speedup!

**Pull request merged**